### PR TITLE
[LI-HOTFIX] Clean up purgatory when leader replica is kicked out of replica list.

### PR DESCRIPTION
**Pull request by kidkun**
In Current implementation,  if a broker receives `StopReplicaRequest` for leader replica it hosts(which can happen in a partition reassignment and new replica list does not include old leader replica) , it does not check (and try to complete) pending delayed operations for this partition, which could result in observable produce delay. 

One example scenario:
Partition reassignment  of partition A  
[leader : B1, replica list : [B1, B2] ] -> [leader : B2, replica list : [B2, B3] ] 

t0: Controller expands the replica set to [B1, B2, B3]

t1: B1 receives produce request on partition A with` acks=all` and `timetout=T(could be a large value)`. B1 puts request into the `DelayedProducePurgatory`.

t2: Controller elects B2 as the new leader and shrinks the replica set fo [B2, B3], send 
     `LeaderAndIsrRequest`  to B2 and B3; send `StopReplicaRequest` to B1.

t3: B1 wait until T and return `TimeoutException` to producer.

If B1 checks for `DelayedProducePurgatory` upon receiving `StopReplicaRequest`, it could fail the produce request sooner with a `KafkaException`.

### CODE REVIEW

File: core/src/main/scala/kafka/cluster/Partition.scala
Comment (jonlee2) [inline core/src/main/scala/kafka/cluster/Partition.scala]: Nit: Looks like delayed fetch and delete records will be cleaned up as well, not just delayed produce requests.

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala
Comment (jonlee2) [inline core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:210]: Is it okay not to create partition/replica?

Comment (kidkun) [reply core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:210]: I think so, as long as we populate the right `PartitionState` in `LeaderAndIsrRequest`, [`ReplicaManager.becomeLeaderOrFollower()`](https://github.com/linkedin/kafka/blob/45b1ec20b64a74d6938ca928052e5db7264a8640/core/src/main/scala/kafka/server/ReplicaManager.scala#L1067) will create a new `Partition` object if it is missing.

**Review by jonlee2 — APPROVED** [review]
Please wait for Patrick's approval as well.

File: core/src/main/scala/kafka/cluster/Partition.scala
Comment (hzxa21) [inline core/src/main/scala/kafka/cluster/Partition.scala]: This function will be called only when `deletePartition` flag is set to true in the StopReplicaRequest and the partition being deleted exists in the `allPartitions` map (See https://github.com/linkedin/kafka/blob/2.0-li/core/src/main/scala/kafka/server/ReplicaManager.scala#L366)

However, during partition reassignment, controller will send out two StopReplicaRequests to the removed replicas in order (See https://github.com/linkedin/kafka/blob/2.0-li/core/src/main/scala/kafka/controller/ControllerChannelManager.scala#L496):
1. _StopReplicaRequest_ (`deletePartition=false`) - in reaction to OfflineReplica state transition
2. _StopReplicaRequest_ (`deletePartition=true`) - in reaction to ReplicaDeletionStarted state transition

When processing the 1st StopReplicaRequest, the broker already removes the partition from `allPartitions` map (See https://github.com/linkedin/kafka/blob/2.0-li/core/src/main/scala/kafka/server/ReplicaManager.scala#L353) so it is guaranteed that `partition.delete()` will not be called during partition reassignment. You can verify it in your test case by sending _StopReplicaRequest_ (`deletePartition=false`) and then _StopReplicaRequest_ (`deletePartition=true`).

I think instead of completing the delated requests in `Partition.scala`, you can complete them in `ReplicaManager.scala` when processing _StopReplicaRequest_.

**Review by hzxa21 — CHANGES_REQUESTED** [review]

**Review by hzxa21 — APPROVED** [review]
LGTM. Thanks for the patch.

**Pull request merged**
